### PR TITLE
InstructorStudentListAjaxPageAction: UnknownFormatConversionException Conversion = 'm' #4739

### LINKED ISSUE #4739: InstructorStudentListAjaxPageAction: UnknownFormatConversionException Conversion = 'm'
Error Message: teammates.ui.controller.ControllerServlet doPost: TEAMMATESLOG|||page|||System Error Report|||true|||Unknown|||Unknown|||abc%manchester.ac.uk@gtempaccount.com|||Unknown|||TEAMMATES (5.65): New System Exception: Conversion = 'm'
Show/Hide Details >>
Error Message

Conversion = 'm'

Actual user abc%manchester.ac.uk@gtempaccount.com
Request Method POST
User Agent Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/38.0.2125.101 Safari/537.36
Request Url https://teammatesv4.appspot.com/page/instructorStudentListAjaxPage
Request Path /page
Request Parameters {courseidx::0, user::abc%manchester.ac.uk@gtempaccount.com, courseid::CHEN30012-30022-30032}
Stack Trace

java.util.UnknownFormatConversionException: Conversion = 'm'
	at java.util.Formatter$FormatSpecifier.conversion(Formatter.java:2646)
	at java.util.Formatter$FormatSpecifier.(Formatter.java:2675)
	at java.util.Formatter.parse(Formatter.java:2528)
	at java.util.Formatter.format(Formatter.java:2469)
	at java.util.Formatter.format(Formatter.java:2423)
	at java.lang.String.format(String.java:2792)
	at teammates.ui.controller.InstructorStudentListAjaxPageAction.execute(InstructorStudentListAjaxPageAction.java:50)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:337)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:143)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:148)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:468)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:256)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)

|||/page/instructorStudentListAjaxPage?courseid=CHEN30012-30022-30032&user=abc%25manchester.ac.uk%40gtempaccount.com&courseidx=0|||ryan.mcgeown%xyz@gtempaccount.com%2016020917322064

### PR DESCRIPTION
**Pull request by thyageshm**
fixes #4739

### CODE REVIEW

**Comment by thyageshm** [comment]
@damithc @wkurniawan07 I can modify one of the instructor's googleIds to test this. should I do it? It would affect a number of html files

**Comment by damithc** [comment]
Extract to a method and unit test the method?

**Comment by thyageshm** [comment]
sir, I have now used a method in the student attributes that was already created for this purpose. Now that only part that is not tested is the addition of the user param, which I believe does not have to be tested

**Comment by damithc** [comment]
Can merge if the reviewer is ok.

File: src/main/java/teammates/ui/controller/InstructorCourseDetailsPageData.java [hunk 1]
```diff
@@ -52,20 +52,20 @@ public void init(InstructorAttributes currentInstructor, CourseDetailsBundle cou
...
         this.sections = new ArrayList<StudentListSectionData>();
+        String studentPhotoUrl = "";
         for (SectionDetailsBundle section: courseDetails.sections) {
             Map<String, String> emailPhotoUrlMapping = new HashMap<String, String>();
             for (TeamDetailsBundle teamDetails : section.teams) {
                 for (StudentAttributes student : teamDetails.students) {
-                    emailPhotoUrlMapping.put(student.email, String.format(photoUrl,
-                                                                          StringHelper.encrypt(student.email),
-                                                                          StringHelper.encrypt(student.course)));
+                    studentPhotoUrl = student.getPublicProfilePictureUrl();
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/InstructorCourseDetailsPageData.java]: Can `studentPhotoUrl` be instantiated here instead? I don't think it's used outside the loop.

File: src/main/java/teammates/ui/controller/InstructorCourseDetailsPageData.java [hunk 2]
```diff
@@ -52,20 +52,20 @@ public void init(InstructorAttributes currentInstructor, CourseDetailsBundle cou
...
+          
+                    // userid is added AFTER the formatting done above to avoid special 
+                    // characters in the userid from affecting the String.format function
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/InstructorCourseDetailsPageData.java]: The comment is no longer relevant, I suppose.

File: src/main/java/teammates/ui/controller/InstructorStudentListAjaxPageAction.java [hunk 3]
```diff
@@ -37,19 +37,19 @@ protected ActionResult execute() throws EntityDoesNotExistException {
...
-                    emailPhotoUrlMapping.put(student.email, String.format(photoUrl,
-                                                                          StringHelper.encrypt(student.email),
-                                                                          StringHelper.encrypt(student.course)));
+                    studentPhotoUrl = student.getPublicProfilePictureUrl();
+                    
+                    // userid is added AFTER the formatting done above to avoid special 
+                    // characters in the userid from affecting the String.format function
+                    studentPhotoUrl = Url.addParamToUrl(studentPhotoUrl, 
+                                                    Const.ParamsNames.USER_ID, account.googleId);
+                    emailPhotoUrlMapping.put(student.email, studentPhotoUrl);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/InstructorStudentListAjaxPageAction.java]: Same issues here.

**Comment by wkurniawan07** [comment]
Some minor changes and it's good to go.

**Comment by thyageshm** [comment]
ok done!

**Comment by wkurniawan07** [comment]
Good to go :+1:

**Pull request merged**